market/mysterium/mysterium_api.go

```diff
@@ -235,12 +238,28 @@ func (mApi *MysteriumAPI) QueryProposals(query ProposalsQuery) ([]market.Service
 	if err != nil {
 		return nil, err
 	}
+	req.Header.Add("If-None-Match", mApi.latestProposalsEtag)
 
-	var proposalsResponse ProposalsResponse
-	err = mApi.http.DoRequestAndParseResponse(req, &proposalsResponse)
+	res, err := mApi.http.Do(req)
 	if err != nil {
 		return nil, err
 	}
+	defer res.Body.Close()
+
+	if res.StatusCode == http.StatusNotModified {
+		return nil, errors.New("node already contains newest proposals")
+	}
+
+	if err := requests.ParseResponseError(res); err != nil {
+		return nil, err
+	}
+
+	var proposalsResponse ProposalsResponse
+	if err := requests.ParseResponseJSON(res, &proposalsResponse); err != nil {
+		return nil, fmt.Errorf("cannot parse proposals response: %v", err)
```

Comment:
Maybe it's better to `errors.Wrap` the error?

Reply:
Yep, will do.